sudospawner/spawner.py

```diff
@@ -94,6 +94,19 @@ def start(self):
 
     @gen.coroutine
     def _signal(self, sig):
+        if sig == 0:
+            # short-circuit existence check without invoking sudo
+            try:
+                os.kill(self.pid, sig)
+            except ProcessLookupError:
+                # No such process
+                return False
+            except PermissionError:
+                # When running hub with reduced permissions,
+                # we won't have permission to send signals, even 0.
+                # PermissionError means process exists.
```

Comment:
"Should be" meaning one should probably be added, but it isn't there right now. This ought to be a rare issue, though, because something has to have gone wrong to get into this state: pid should only be stored while the Hub believes the server is running. So if the pid points to another process, it means that the server stopped and another process claimed the same pid *while the Hub thought the server was running*. The most likely cause of this is an unclean shutdown of the Hub, where pid state remains in the database, and then reboot / etc. In general, when shutting down the Hub and cleaning up all single-user servers, there should be no PIDs in the database at all. It's also possible there's a bug somewhere causing this state to not be cleared correctly.